### PR TITLE
Add AddressDecoder Documentation and Test.

### PR DESCRIPTION
**Pull request by sequencer**
**Type of change**: other enhancement

**Impact**: no functional change
small API change to make 
```
object AddressDecoder {
-  val debug = false
+ var debug = true
}
```
to let user debug it without changing source code.
change `import Chisel._` to `import chisel3._`

**Development Phase**: implementation

**Release Notes**
Add documentation to `AddressDecoder`.

### CODE REVIEW

File: src/main/scala/diplomacy/AddressDecoder.scala [hunk 1]
```diff
@@ -2,111 +2,164 @@
...
   def bitScore(partitions: Partitions): Seq[Int] = {
+    /** Detect partition change. */
     val maxPortsPerPartition = partitions.map(_.size).max
+    /** Detect port change. */
     val maxSetsPerPartition = partitions.map(_.map(_.size).sum).max
+    /** Detect partition change. */
     val sumSquarePortsPerPartition = partitions.map(p => p.size * p.size).sum
+    /** Detect port change. */
     val sumSquareSetsPerPartition = partitions.map(_.map(p => p.size * p.size).sum).max
     Seq(maxPortsPerPartition, maxSetsPerPartition, sumSquarePortsPerPartition, sumSquareSetsPerPartition)
```
Comment (sequencer) [inline src/main/scala/diplomacy/AddressDecoder.scala:105]: After understanding the partition algorithm.
I think `sumSquarePortsPerPartition` should be priority to `maxSetsPerPartition`?
What's your think @aswaterman @terpstra

File: src/main/scala/diplomacy/AddressDecoder.scala [hunk 2]
```diff
@@ -2,111 +2,164 @@
...
+
+  /** cases already distinguished by the selected bits thus far. */
   type Partitions = Seq[Partition]
 
+  /* require the port to be ordered, since we need to adjoin same addresses. */
   val addressOrder = Ordering.ordered[AddressSet]
   val portOrder = Ordering.Iterable(addressOrder)
   val partitionOrder = Ordering.Iterable(portOrder)
 
-  // Find the minimum subset of bits needed to disambiguate port addresses.
```
Comment (mwachs5) [inline src/main/scala/diplomacy/AddressDecoder.scala:18]: why remove this description? It seems like they are still helpful...

File: src/main/scala/diplomacy/AddressDecoder.scala [hunk 3]
```diff
@@ -2,111 +2,164 @@
...
+    */
   def apply(ports: Ports, givenBits: BigInt = BigInt(0)): BigInt = {
-    val nonEmptyPorts = ports.filter(_.nonEmpty)
+    def checkOverlap(ports: Ports): Unit = ports.combinations(2).foreach { case Seq(x, y) =>
+      x.foreach { a => y.foreach { b =>
+        require (!a.overlaps(b), s"Ports cannot overlap: $a $b")
+      } }
+    }
+
+    /* remove all empty ports. */
```
Comment (mwachs5) [inline src/main/scala/diplomacy/AddressDecoder.scala:53]:
```suggestion
    // remove all empty ports.
```

File: src/main/scala/diplomacy/AddressDecoder.scala [hunk 4]
```diff
@@ -2,111 +2,164 @@
...
     if (nonEmptyPorts.size <= 1) {
       givenBits
     } else {
-      // Verify the user did not give us an impossible problem
-      nonEmptyPorts.combinations(2).foreach { case Seq(x, y) =>
-        x.foreach { a => y.foreach { b =>
-          require (!a.overlaps(b), s"Ports cannot overlap: $a $b")
-        } }
-      }
+      /* Verify the user did not give us an impossible problem. */
```
Comment (mwachs5) [inline src/main/scala/diplomacy/AddressDecoder.scala:58]:
```suggestion
      // Verify the user did not give us an impossible problem.
```

File: src/main/scala/diplomacy/AddressDecoder.scala [hunk 5]
```diff
@@ -2,111 +2,164 @@
...
   }
 
-  // The algorithm has a set of partitions, discriminated by the selected bits.
-  // Each partion has a set of ports, listing all addresses that lead to that port.
-  // Seq[Seq[Seq[AddressSet]]]
-  //         ^^^^^^^^^^^^^^^ set of addresses that are routed out this port
-  //     ^^^ the list of ports
-  // ^^^ cases already distinguished by the selected bits thus far
-  //
-  // Solving this problem is NP-hard, so we use a simple greedy heuristic:
```
Comment (mwachs5) [inline src/main/scala/diplomacy/AddressDecoder.scala:66]: why did you remove this bit of explanation?

Comment (jerryhethatday) [reply src/main/scala/diplomacy/AddressDecoder.scala:66]: I felt like these comments are not in the right place.

**Comment by sequencer** [comment]
essentially the AddressDecoder should be replaced by BitSet.decoder from chisel3.
I'm closing this and in the future a new AddressDecoder should be added to RC.
